docs/docs/Best_Practices/Message_Limits.md

```diff
@@ -0,0 +1,27 @@
+# Message Limits
+
+There are limits on how much data can be sent back and forth between the plugin and engine at a time. There are five scenarios where this comes into play:
+1. Inputs sent from plugin to engine, as arguments to a [Plugin Operation](/References/Plugin_Operations.md). For example, the schema-defined `Repository` object that is provided as input to plugin operations.
+2. Outputs sent back from the plugin to the engine, as the return values from plugin operations.
+3. Exception messages and call stacks thrown by plugin code. For example, the `message` field within [User Visible Errors](/Best_Practices/User_Visible_Errors.md).
+4. Inputs sent from plugin to engine, as arguments to a [Platform library](/References/Platform_Libraries.md) function. For example, the `message` field that is passed to `logger.debug`. 
+5. Outputs sent back from the plugin to the engine, as the return values from Platform Library functions. For example, the `stdout` resulting from a call to `libs.run_bash`. 
+
+For case 1 and 2, the total size of data must be less than 4 mebibytes (4 MiB).
+
+For case 3, the total size of data must be less than 128 kibibytes (128 KiB).
+
+For case 4 and 5, the total size of data must be less than 32 mebibytes (32 MiB).
```

Comment:
should we change the code and make (1) and (2) the same as (4) and (5)?

Reply:
We haven't changed the code yet, mostly because we haven't noticed any issues with the plugins we use for testing, but I could imagine we would hit this limit when our plugins created a Repository schema with a million properties.

It's important to note that we only have tunables for the metadata on the DockerPluginRuntime side and for the message on the LibsServer side. We wouldn't be able to configure the limits for message on DockerPluginRuntime or metadata on LibsServer. We should probably (at the very least) make these tunables.

Reply:
Lets file tracking bugs for these? Would be nice to make everything consistent at 32 MiB